README.md
# piveau pipe connector

## Features
 * Asynchronous event based pipe handling
 * piveau conformable logging
 * Internal management of pipe structure
 * Convenient API
 * Vert.x support

## Get started

The most simple example would be:
```java
PipeConnector.create().handle(pipeContext -> pipeContext.setResult("Hello World!"));
```

Of course, a more practical example would be this:
```java
PipeConnector connector = PipeConnector.create();
connector.handle(pipeContext -> {
    JsonNode config = pipeContext.getConfig();
    String data = pipeContext.getStringData();

    // do your job here
    String result = ...
    
    pipeContext.setResult(result); 
});
```

If you're already in a Vert.x context and would like to use a worker Verticle with an event bus address, this will do:
```java
PipeConnector.create(vertx, ar -> {
    if (ar.succeeded()) {
        ar.result().consumerAddress(EVENTBUS_ADDRESS);
        startFuture.complete();
    } else {
        startFuture.fail(ar.cause());
    }
});
```

## Service Interface

* `/pipe`

    Pipe endpoint (see ReDoc on root context path)

* `/health`

    Returns up status and build infos. Requires a buildInfo.json in the classpath, e.g.:
    ```json
    {
       "timestamp": "${buildTimestamp}",
       "version": "${project.version}"
    }
    ```
    `timestamp` and `version` could be set automatically by maven. 

* `/metrics`

    Internal metric infos

It is possible to mount a sub-router for your own interfaces:

```java
connector.subRouter("/mountPoint", router);
```

## Configuration

 * PIVEAU_PIPE_ENDPOINT_PORT - Default port is 8080
 * PIVEAU_PIPE_MAIL_CONFIG - JSON object:
   
   ```json
   {
      "host": "<smtp_host>",
      "system": "<environment>",
      "mailto": "<receiver_address>"
   }
   ```
  

CHANGELOG.md
# ChangeLog

## [2.0.3](https://gitlab.fokus.fraunhofer.de/viaduct/pipe-connector/tags/2.0.3) (2019-10-16)

**Added:**
* Possibility to mount a sub-router to the connector

## [2.0.2](https://gitlab.fokus.fraunhofer.de/viaduct/pipe-connector/tags/2.0.2) (2019-10-02)

**Fixed:**
* Failed check negated

## [2.0.1](https://gitlab.fokus.fraunhofer.de/viaduct/pipe-connector/tags/2.0.1) (2019-08-28)

**Changed**
* Requires now latest LTS Java 11
* Update pipe model to 2.1.0

## [2.0.0](https://gitlab.fokus.fraunhofer.de/viaduct/pipe-connector/tags/2.0.0) (2019-08-22)

**Changed**
* `PipeLogger`, `PipeContext` and `PipeMailer` classes implemented in kotlin
* Updated to pipe-model version 2.0.0. This forces a major version update as well!

## [1.0.4](https://gitlab.fokus.fraunhofer.de/viaduct/pipe-connector/tags/1.0.4) (2019-08-13)

**Fixed:**
* Another hotfix for deep copy of pipe.

## [1.0.3](https://gitlab.fokus.fraunhofer.de/viaduct/pipe-connector/tags/1.0.3) (2019-08-12)

**Fixed:**
* Hotfix for deep copy of pipe.

## [1.0.2](https://gitlab.fokus.fraunhofer.de/viaduct/pipe-connector/tags/1.0.2) (2019-08-12)

**Fixed:**
* `setPayloadData` passing non null value of `DataType` (kotlin)

## [1.0.1](https://gitlab.fokus.fraunhofer.de/viaduct/pipe-connector/tags/1.0.1) (2019-08-09)

**Fixed:**
* Update pipe-model, fix for Pipe deserialization

## [1.0.0](https://gitlab.fokus.fraunhofer.de/viaduct/pipe-connector/tags/1.0.0) (2019-08-09)

**Added:**
* kotlin support
* Test stage as part of gitlab ci

**Changed:**
* Upgrade to Vert.x 3.8.0
* Upgrade to kotlin 1.3.41 (no implementations yet)
* Upgrade to pipe-model 1.0.0 (pure kotlin)
 
## [0.0.4](https://gitlab.fokus.fraunhofer.de/viaduct/pipe-connector/tags/0.0.4) (2019-??-??)

**Added:**
* `/health` path up status with build info
* `/metrics` path with metrics info
* `/config/schema` returns config json schema

**Changed:**
* Log output in case of forwarding
* Upgrade to Vert.x 3.7.1

**Fixed:**
* Resource filtering before build

## [0.0.3](https://gitlab.fokus.fraunhofer.de/viaduct/pipe-connector/tags/0.0.3) (2019-04-15)

**Added:**
- EMail support
- Configuration for the release plugin in the pom file
- ChangeLog

**Changed:**
- Upgrade to Vert.x 3.7.0

## [0.0.2](https://gitlab.fokus.fraunhofer.de/viaduct/pipe-connector/tags/pipe-connector-0.0.2) (2019-03-01)

## [0.0.1](https://gitlab.fokus.fraunhofer.de/viaduct/pipe-connector/tags/pipe-connector-0.0.1) (2019-03-01)


src/test/java/io/piveau/pipe/connector/PipeLoggerTest.java
package io.piveau.pipe.connector;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.piveau.pipe.Pipe;
import io.piveau.pipe.PipeLogger;
import io.vertx.core.Vertx;
import io.vertx.junit5.VertxExtension;
import io.vertx.junit5.VertxTestContext;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import java.io.IOException;

@DisplayName("Testing the pipe connector")
@ExtendWith(VertxExtension.class)
public class PipeLoggerTest {

    @Test
    @DisplayName("Testing handler")
    public void logTest(Vertx vertx, VertxTestContext testContext) {
        vertx.fileSystem().readFile("test-pipe.json", result -> {
            if (result.succeeded()) {
                try {
                    Pipe pipe = new ObjectMapper().readValue(result.result().toString(), Pipe.class);
                    PipeLogger logger = new PipeLogger(pipe.getHeader(), pipe.getBody().getSegments().get(0).getHeader());
                    logger.info("Test");
                    logger.info("Test {}", "param1");
                    logger.info("Test {}, {}", "param1", "param2");
                    logger.info("Test {}, {}, {}", "param1", "param2", "param3");
                    logger.info("Test", new Throwable("throwable1"));
                    testContext.completeNow();
                } catch (IOException e) {
                    testContext.failNow(e);
                }
            }
        });
    }

}


src/test/java/io/piveau/pipe/connector/PipeConnectorTest.java
package io.piveau.pipe.connector;

import io.vertx.core.Vertx;
import io.vertx.core.buffer.Buffer;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.web.client.WebClient;
import io.vertx.junit5.Checkpoint;
import io.vertx.junit5.VertxExtension;
import io.vertx.junit5.VertxTestContext;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

@DisplayName("Testing the pipe connector")
@ExtendWith(VertxExtension.class)
public class PipeConnectorTest {

    static PipeConnector connector;

    @BeforeAll
    @DisplayName("Initialize connector")
    public static void createConnector(Vertx vertx, VertxTestContext testContext) {

        PipeConnector.create(vertx, ar -> {
            if (ar.succeeded()) {
                connector = ar.result();
                testContext.completeNow();
            } else {
                testContext.failNow(ar.cause());
            }
        });

    }

    @Test
    @DisplayName("Testing handler")
    public void testHandler(Vertx vertx, VertxTestContext testContext) {
        Checkpoint checkpoint = testContext.checkpoint();

        connector.handler(pipeContext -> {
            System.out.println(pipeContext.getPipe().getHeader().getTitle());
            checkpoint.flag();
        });

        webInjection(vertx, testContext, "test-pipe.json");
    }

    @Test
    @DisplayName("Testing producer")
    public void testProducer(Vertx vertx, VertxTestContext testContext) {
        Checkpoint checkpoint = testContext.checkpoint();
        vertx.eventBus().consumer("io.piveau.pipe.connector.testing", msg -> {
            PipeContext pipeContext = (PipeContext)msg.body();
            System.out.println(pipeContext.getPipe().getHeader().getTitle());
            checkpoint.flag();
            msg.reply("success");
        });

        connector.consumerAddress("io.piveau.pipe.connector.testing");

        webInjection(vertx, testContext, "test-pipe.json");
    }

    private void webInjection(Vertx vertx, VertxTestContext testContext, String fileName) {
        Buffer buffer = vertx.fileSystem().readFileBlocking(fileName);
        JsonObject pipe = new JsonObject(buffer);
        WebClient client = WebClient.create(vertx);
        client.post(8080, "localhost", "/pipe")
                .putHeader("content-type", "application/json")
                .sendJsonObject(pipe, response -> {
                    if (response.succeeded()) {
                        if (response.result().statusCode() == 202) {
                            testContext.completeNow();
                        } else {
                            testContext.failNow(new Throwable(response.result().statusMessage()));
                        }
                    } else {
                        testContext.failNow(response.cause());
                    }
                });
    }

}

src/main/kotlin/io.piveau.pipe/PipeLogger.kt
package io.piveau.pipe

import org.slf4j.Logger
import org.slf4j.LoggerFactory

class PipeLogger(pipeHeader: PipeHeader, segmentHeader: SegmentHeader) {
    private val log: Logger = LoggerFactory.getLogger("pipe.${pipeHeader.id}")
    private val prefix: String = "${pipeHeader.runId} [${pipeHeader.title} (${pipeHeader.context})] [${segmentHeader.title} (${segmentHeader.name})] "

    fun error(message: String?) = log.error(prefix + message)

    fun error(format: String?, argument1: Any?) = log.error(prefix + format, argument1)

    fun error(format: String?, argument1: Any?, argument2: Any?) = log.error(prefix + format, argument1, argument2)

    fun error(format: String?, vararg arguments: Any?) = log.error(prefix + format, *arguments)

    fun error(format: String?, e: Throwable?) = log.error(prefix + format, e)

    fun warn(message: String?) = log.warn(prefix + message)

    fun warn(format: String, argument1: Any) = log.warn(prefix + format, argument1)

    fun warn(format: String, argument1: Any, argument2: Any) = log.warn(prefix + format, argument1, argument2)

    fun warn(format: String, vararg arguments: Any) = log.warn(prefix + format, *arguments)

    fun warn(format: String, e: Throwable) = log.warn(prefix + format, e)

    fun info(message: String) = log.info(prefix + message)

    fun info(format: String, argument1: Any) = log.info(prefix + format, argument1)

    fun info(format: String, argument1: Any, argument2: Any) = log.info(prefix + format, argument1, argument2)

    fun info(format: String, vararg arguments: Any) = log.info(prefix + format, *arguments)

    fun info(format: String, e: Throwable) = log.info(prefix + format, e)

    fun debug(message: String) = log.debug(prefix + message)

    fun debug(format: String, argument1: Any) = log.debug(prefix + format, argument1)

    fun debug(format: String, argument1: Any, argument2: Any) = log.debug(prefix + format, argument1, argument2)

    fun debug(format: String, vararg arguments: Any) = log.debug(prefix + format, *arguments)

    fun debug(format: String, e: Throwable) = log.debug(prefix + format, e)

    fun trace(message: String) = log.trace(prefix + message)

    fun trace(format: String, argument1: Any) = log.trace(prefix + format, argument1)

    fun trace(format: String, argument1: Any, argument2: Any) = log.trace(prefix + format, argument1, argument2)

    fun trace(format: String, vararg arguments: Any) = log.trace(prefix + format, *arguments)

    fun trace(format: String, e: Throwable) = log.trace(prefix + format, e)

}

src/main/kotlin/io/piveau/pipe/connector/PipeMailer.kt
package io.piveau.pipe.connector

import io.vertx.core.Vertx
import io.vertx.core.buffer.Buffer
import io.vertx.core.json.JsonObject
import io.vertx.ext.mail.MailAttachment
import io.vertx.ext.mail.MailClient
import io.vertx.ext.mail.MailConfig
import io.vertx.ext.mail.MailMessage
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.thymeleaf.TemplateEngine
import org.thymeleaf.context.Context
import org.thymeleaf.templatemode.TemplateMode
import org.thymeleaf.templateresolver.ClassLoaderTemplateResolver

class PipeMailer(vertx: Vertx, val config: JsonObject) {

    companion object {
        @JvmStatic
        fun create(vertx: Vertx, config: JsonObject): PipeMailer = PipeMailer(vertx, config)
    }

    private val log: Logger? = LoggerFactory.getLogger(this.javaClass)

    private val client: MailClient?

    private val logo: Buffer

    private val templateEngine: TemplateEngine

    init {
        val mailConfig = MailConfig(config.getString("host"))
        mailConfig.isTrustAll = true

        client = MailClient.createShared(vertx, mailConfig)

        logo = vertx.fileSystem().readFileBlocking("emails/piveau_logo.png")

        val templateResolver = ClassLoaderTemplateResolver().also {
            it.templateMode = TemplateMode.HTML
            it.prefix = "/emails/"
            it.suffix = ".html"
        }

        templateEngine = TemplateEngine()
        templateEngine.setTemplateResolver(templateResolver)
    }

    fun send(pipeContext: PipeContext) {
        val pipe = pipeContext.pipeManager.pipe
        val segment = pipeContext.pipeManager.currentSegment
        val address = pipeContext.pipeManager.config.path("mailto")?.asText(config.getString("mailto", "piveau-support@fokus.fraunhofer.de"))

        val attachment = MailAttachment()
        with(attachment) {
            contentType = "image/png"
            data = logo
            disposition = "inline"
            contentId = "logo"
        }

        val ctx = Context()
        ctx.setVariables(mapOf(
                "logo" to "logo",
                "header" to segment?.header?.title,
                "harvester" to pipe.header.title,
                "message" to pipeContext.cause?.message,
                "details" to JsonObject(segment?.body?.config.toString()).encodePrettily(),
                "run" to pipe.header.startTime
        ))

        val message = MailMessage()
        with(message) {
            to = listOf(address)
            from = "noreply@piveau.io"
            subject = "[${pipe.header.context} - ${config.getString("system", "consus")}] [${segment?.header?.name}] piveau event"
            inlineAttachment = listOf(attachment)
            html = templateEngine.process("simple", ctx)
        }

        client?.sendMail(message) { ar ->
            if (ar.succeeded()) {
                log?.debug("Mail sent: {}", ar.result().messageID)!!
            } else {
                log?.error("Mail sent", ar.cause())!!
            }
        }!!
    }

}


src/main/kotlin/io/piveau/pipe/connector/PipeContext.kt
package io.piveau.pipe.connector

import com.fasterxml.jackson.databind.node.ObjectNode
import io.piveau.pipe.*
import io.vertx.core.Vertx
import io.vertx.core.buffer.Buffer
import io.vertx.core.http.HttpMethod
import io.vertx.core.json.Json
import io.vertx.core.json.JsonObject
import io.vertx.ext.web.client.WebClient
import java.net.URL

class PipeContext private constructor(pipe: Pipe, private val pipeConnector: PipeConnector? = null) {

    val pipeManager = PipeManager.create(pipe)

    private val pipeLogger = PipeLogger(pipe.header, pipeManager.currentSegment!!.header)

    constructor(
        jsonObject: JsonObject,
        pipeConnector: PipeConnector? = null
    ) : this(Json.mapper.convertValue(jsonObject, Pipe::class.java), pipeConnector)

    constructor(buffer: Buffer, pipeConnector: PipeConnector? = null) : this(
        Json.decodeValue<Pipe>(
            buffer,
            Pipe::class.java
        ), pipeConnector
    )

    var forwarded: Boolean = false

    fun isForwarded(): Boolean = forwarded

    private var failureThrowable: Throwable? = null

    val failure: Boolean
        get() = failureThrowable != null

    fun isFailure(): Boolean = failure

    val cause: Throwable?
        get() = failureThrowable

    private var textResult: String? = null
    private var byteResult: ByteArray? = null

    private var mimeTypeResult: String? = null

    private var dataInfoResult: ObjectNode? = null

    val config = pipeManager.config

    val mimeType: String?
        get() = pipeManager.mimeType

    val dataInfo: ObjectNode
        get() = pipeManager.dataInfo

    val pipe: Pipe
        get() = pipeManager.pipe

    val log: PipeLogger
        get() = pipeLogger

    fun log() = pipeLogger

    val nextEndpoint: Endpoint?
        get() = pipeManager.nextEndpoint

    val stringData: String
        get() = pipeManager.data

    val binaryData: ByteArray
        get() = pipeManager.binaryData

    val message: JsonObject
        get() {
            val message = JsonObject().put("config", pipeManager.config)
            if (pipeManager.isBase64Payload) {
                message.put("data", pipeManager.binaryData).put("dataType", DataType.base64)
            } else {
                message.put("data", pipeManager.data).put("dataType", DataType.text)
            }
            return message
        }

    fun setResult(result: String, dataMimeType: String? = null, info: ObjectNode? = null): PipeContext {
        textResult = result
        mimeTypeResult = dataMimeType
        dataInfoResult = info
        forwarded = false
        return this
    }

    fun setResult(result: ByteArray, dataMimeType: String? = null, info: ObjectNode? = null): PipeContext {
        byteResult = result
        mimeTypeResult = dataMimeType
        dataInfoResult = info
        forwarded = false
        return this
    }

    fun getFinalPipe(): Buffer? {
        textResult?.let {
            pipeManager.setPayloadData(it, DataType.text, mimeTypeResult, dataInfoResult)
        } ?: byteResult?.let {
            pipeManager.setPayloadData(it, mimeTypeResult, dataInfoResult)
        }
        return Json.encodeToBuffer(pipeManager.getProcessedPipe())
    }

    fun forward() = pipeConnector?.forward(this)

    fun forward(vertx: Vertx) = forward(WebClient.create(vertx))

    fun forward(webClient: WebClient) {
        if (forwarded) {
            log.warn("Already forwarded")
            return
        }

        nextEndpoint?.let { endpoint ->
            getFinalPipe().let { pipe ->
                val address = URL(endpoint.address)
                val method = HttpMethod.valueOf(endpoint.method ?: "POST")
                webClient.request(method, address.port, address.host, address.path)
                    .putHeader("Content-Type", "application/json")
                    .sendBuffer(pipe) {
                        if (it.succeeded()) {
                            if (it.result().statusCode() == 200 || it.result().statusCode() == 202) {
                                log.trace("successfully forwarded")
                            } else {
                                log.error("{} - {}", it.result().statusCode(), it.result().statusMessage())
                            }
                        } else {
                            log.error("Forward request", it.cause())
                        }
                    }
            }
        } ?: log.trace("No next pipe segment available")

        forwarded = true
    }

    fun pass() = pipeConnector?.pass(this)

    fun pass(vertx: Vertx) = pass(WebClient.create(vertx))

    fun pass(webClient: WebClient) {
        if (forwarded) {
            log.warn("Already forwarded")
            return
        }

        if (pipeManager.isBase64Payload) {
            setResult(pipeManager.binaryData, pipeManager.mimeType, pipeManager.dataInfo)
        } else {
            setResult(pipeManager.data, pipeManager.mimeType, pipeManager.dataInfo)
        }
        forward(webClient)
    }

    fun setFailure(message: String) = setFailure(Throwable(message))

    fun setFailure(throwable: Throwable): PipeContext {
        failureThrowable = throwable

        log.error("Pipe failure: " + throwable.message, throwable.cause ?: throwable)

        if (pipeConnector?.isMailerEnabled == true) {
            pipeConnector.useMailer().send(this)
        }
        return this
    }

}


src/main/java/io/piveau/pipe/connector/PipeContextMessageCodec.java
package io.piveau.pipe.connector;

import io.vertx.core.buffer.Buffer;
import io.vertx.core.eventbus.MessageCodec;
import io.vertx.core.json.JsonObject;

class PipeContextMessageCodec implements MessageCodec<JsonObject, PipeContext> {

    private PipeConnector pipeConnector;

    PipeContextMessageCodec(PipeConnector pipeConnector) {
        this.pipeConnector = pipeConnector;
    }

    @Override
    public void encodeToWire(Buffer buffer, JsonObject entries) {

    }

    @Override
    public PipeContext decodeFromWire(int i, Buffer buffer) {
        return null;
    }

    @Override
    public PipeContext transform(JsonObject entries) {
        return new PipeContext(entries, pipeConnector);
    }

    @Override
    public String name() {
        return getClass().getSimpleName();
    }

    @Override
    public byte systemCodecID() {
        return -1;
    }

}


src/main/java/io/piveau/pipe/connector/PipeConnector.java
package io.piveau.pipe.connector;

import io.vertx.config.ConfigRetriever;
import io.vertx.config.ConfigRetrieverOptions;
import io.vertx.config.ConfigStoreOptions;
import io.vertx.core.*;
import io.vertx.core.eventbus.DeliveryOptions;
import io.vertx.core.http.HttpMethod;
import io.vertx.core.http.HttpServer;
import io.vertx.core.http.HttpServerOptions;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.dropwizard.MetricsService;
import io.vertx.ext.healthchecks.HealthCheckHandler;
import io.vertx.ext.healthchecks.Status;
import io.vertx.ext.web.Router;
import io.vertx.ext.web.RoutingContext;
import io.vertx.ext.web.api.contract.RouterFactoryOptions;
import io.vertx.ext.web.api.contract.openapi3.OpenAPI3RouterFactory;
import io.vertx.ext.web.handler.CorsHandler;
import io.vertx.ext.web.handler.StaticHandler;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.Date;
import java.util.UUID;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class PipeConnector extends AbstractVerticle {

    private Logger log = LoggerFactory.getLogger(getClass());

    private Handler<PipeContext> handler;

    private String consumer;

    private PipeMailer pipeMailer;

    private Router router;

    public static PipeConnector create() {
        return create(new DeploymentOptions());
    }

    public static PipeConnector create(DeploymentOptions deploymentOptions) {
        Vertx vertx = Vertx.vertx();

        PipeConnector pipeConnector = new PipeConnector();
        vertx.deployVerticle(pipeConnector, deploymentOptions);
        vertx.eventBus().registerCodec(new PipeContextMessageCodec(pipeConnector));
        return pipeConnector;
    }

    public static void create(Vertx vertx, Handler<AsyncResult<PipeConnector>> handler) {
        create(vertx, new DeploymentOptions(), handler);
    }

    public static void create(Vertx vertx, DeploymentOptions deploymentOptions, Handler<AsyncResult<PipeConnector>> handler) {
        PipeConnector pipeConnector = new PipeConnector();
        vertx.eventBus().registerCodec(new PipeContextMessageCodec(pipeConnector));
        vertx.deployVerticle(pipeConnector, deploymentOptions, ar -> {
            if (ar.succeeded()) {
                handler.handle(Future.succeededFuture(pipeConnector));
            } else {
                handler.handle(Future.failedFuture(ar.cause()));
            }
        });
    }

    @Override
    public void start(Promise<Void> startPromise) {
        MetricsService metrics = MetricsService.create(vertx);

        ConfigStoreOptions storeOptions = new ConfigStoreOptions()
                .setType("env")
                .setConfig(new JsonObject().put("keys", new JsonArray()
                        .add("PIVEAU_PIPE_ENDPOINT_PORT")
                        .add("PIVEAU_PIPE_MAIL_CONFIG")));

        ConfigRetriever retriever = ConfigRetriever.create(vertx, new ConfigRetrieverOptions().addStore(storeOptions));
        retriever.getConfig(lc -> {
            if (lc.succeeded()) {
                JsonObject config = lc.result();
                if (config.containsKey("PIVEAU_PIPE_MAIL_CONFIG")) {
                    JsonObject mailConfig = config.getJsonObject("PIVEAU_PIPE_MAIL_CONFIG");
                    pipeMailer = PipeMailer.create(vertx, mailConfig);
                }
                OpenAPI3RouterFactory.create(vertx, "webroot/openapi-pipe.yaml", ar -> {
                    if (ar.succeeded()) {

                        OpenAPI3RouterFactory routerFactory = ar.result();
                        RouterFactoryOptions options = new RouterFactoryOptions().setMountNotImplementedHandler(true);
                        routerFactory.setOptions(options);
                        routerFactory.addHandlerByOperationId("incomingPipe", this::incomingPipeHandler);
                        routerFactory.addHandlerByOperationId("configSchema", this::configSchemaHandler);

                        router = routerFactory.getRouter();
                        router.route().order(0).handler(CorsHandler.create("*").allowedMethods(Stream.of(HttpMethod.POST).collect(Collectors.toSet())));

                        router.route("/*").handler(StaticHandler.create());

                        HealthCheckHandler hch = HealthCheckHandler.create(vertx);
                        hch.register("buildInfo", future -> vertx.fileSystem().readFile("buildInfo.json", bi -> {
                            if (bi.succeeded()) {
                                future.complete(Status.OK(bi.result().toJsonObject()));
                            } else {
                                future.fail(bi.cause());
                            }
                        }));
                        router.get("/health").handler(hch);

                        router.get("/metrics").produces("application/json").handler(routingContext -> {
                            JsonObject m = metrics.getMetricsSnapshot(vertx);
                            if (m != null) {
                                routingContext.response().setStatusCode(200).end(m.encodePrettily());
                            } else {
                                routingContext.response().setStatusCode(503).end();
                            }
                        });

                        HttpServer server = vertx.createHttpServer(new HttpServerOptions().setPort(config.getInteger("PIVEAU_PIPE_ENDPOINT_PORT", 8080)));
                        server.requestHandler(router).listen();

                        startPromise.complete();
                    } else {
                        startPromise.fail(ar.cause());
                    }
                });
            }
        });
    }

    public void handler(Handler<PipeContext> handler) {
        this.handler = handler;
        this.consumer = null;
    }

    public void consumerAddress(String consumer) {
        this.consumer = consumer;
        this.handler = null;
    }

    public void subRouter(String mountPoint, Router router) {
        if (router != null) {
            this.router.mountSubRouter(mountPoint, router);
        } else {
            throw new IllegalStateException("Router not set.");
        }
    }

    public void incomingPipeHandler(RoutingContext routingContext) {
        JsonObject pipe = routingContext.getBodyAsJson();
        if (!pipe.getJsonObject("header").containsKey("runId")) {
            pipe.getJsonObject("header").put("runId", UUID.randomUUID().toString());
        }
        if (!pipe.getJsonObject("header").containsKey("startTime")) {
            pipe.getJsonObject("header").put("startTime", new Date().toInstant().atZone(ZoneId.systemDefault()).format(DateTimeFormatter.ISO_OFFSET_DATE_TIME));
        }
        if (consumer != null) {
            vertx.eventBus().send(consumer, pipe, new DeliveryOptions().setCodecName(new PipeContextMessageCodec(this).name()));
            routingContext.response().setStatusCode(202).end(pipe.getJsonObject("header").getString("runId"));
        } else if (handler != null) {
            PipeContext pipeContext = new PipeContext(pipe, this);

            vertx.executeBlocking(future -> {
                handler.handle(pipeContext);
                if (pipeContext.isFailure()) {
                    future.fail(pipeContext.getCause());
                } else {
                    if (!pipeContext.isForwarded()) {
                        pipeContext.forward(vertx);
                    }
                    future.complete();
                }
            }, result -> {
                if (result.succeeded()) {
                    pipeContext.log().trace("Incoming pipe successfully handled by handler.");
                } else {
                    pipeContext.log().error("Handling pipe", result.cause());
                }
            });
            routingContext.response().setStatusCode(202).end(pipe.getJsonObject("header").getString("runId"));
        } else {
            routingContext.response().setStatusCode(501).end();
        }

    }

    public void configSchemaHandler(RoutingContext routingContext) {
        vertx.fileSystem().readFile("config.schema.json", ar -> {
            if (ar.succeeded()) {
                routingContext.response().putHeader("Content-Type", "application/schema+json").end(ar.result());
            } else {
                routingContext.response().setStatusCode(404).end();
            }
        });
    }

    public void forward(PipeContext pipeContext) {
        if (!pipeContext.isForwarded()) {
            pipeContext.forward(vertx);
        }
    }

    public void pass(PipeContext pipeContext) {
        if (!pipeContext.isForwarded()) {
            pipeContext.pass(vertx);
        }
    }

    public boolean isMailerEnabled() {
        return pipeMailer != null;
    }

    public PipeMailer useMailer() {
        return pipeMailer;
    }

}
